### PR TITLE
CSS refactoring focused on performance improvements and accountability

### PR DESCRIPTION
**Pull request by PanayotCankov**
Improves node-tests, now named unit-tests, new npm scripts:
npm run unit-test
npm run unit-test-watch
The unit-test compilation tests only functionality shared for all platforms and as such does not involve platform d.ts-es.
npm run unit-test-watch relies on mocha-typescript, it will rebuild TSC (in ~3sec on my mac mini) and re-run mocha tests.

Added VSCode launching script for the unit-test so they can be launched and debugged from VSCode.

Moved to node tests for easier maintenance and faster execution:
 - css selector parser tests
 - css matching tests

Deleted node-tests/definitions, added dev dependencies to @types/chai etc. Updated tsconfig.*.json files types field to include such dependencies only in the appropriate contexts.

Updated the TypeScript version to ^2.5.2, everything seemed fine.

Added an internal "loadAppCss" event on application so style-scope can subscribe for it instead of the more generic "launch". Styles now can be loaded earlier than "launch" by calling application.loadAppCss(). Example `vendor.js`:
```
// Snapshot the ~/app.css and the theme
import * as application from "application";
import "ui/styling/style-scope"; // When required, wires for application events.
global.registerModule("app.css", () => require("~/app.css"));
global.registerModule("~nativescript-theme-core/css/core.light.css", () => require("nativescript-theme-core/css/core.light.css"));
application.loadAppCss();

require("./vendor-platform");
require("bundle-entry-points");
```

Implemented the basics of a CSS3 spec parser
CSS Syntax Module Level 3 https://www.w3.org/TR/css-syntax-3/
It is somewhat 2-3 times faster than the on by rework.
There are some unhandled branches in the implementation, so it is hidden behind a feature toggle. To enable it in an app, set in the `app/package.json`:
```
{
    "cssParser": "nativescript"
}
```

Changed object[privatePropertyKey] to object._privateProperty for the sake of sanity.

Added a `global.moduleResolvers` array consisting of the `global.require` function.

Fixed a problem with latest npm versions, that `npm pack` will not collect module's internal package.json files.

Fixed the `@profile` to work on static methods with the `{ "profiling": "timeline" }`.

Refactored the selectors parser to move the selectors ast closer to the selectors spec.

Added `@profile` decorators on various methods on the CSSSource class so resolving, fs loading and parsing can be separately tracked per css file.

The style-scope has been refactored in a way it can now work with modules that export css string, object that "toString()" will produce css or JSON with CSS AST. This enables it to work with web pack and consume "css-loader" directly as well as the experimental "nativescript-dev-webpack/css2json-loader". We have to further record times in various apps.

### CODE REVIEW

File: tns-core-modules/ui/styling/style-scope.ts [hunk 1]
```diff
@@ -179,15 +248,22 @@ const loadCss = profile(`"style-scope".loadCss`, (cssFile: string) => {
...
+let loadedAppCSS = null;
+export const loadAppCSS = profile('"style-scope".loadAppCSS', (args: application.LoadAppCSSEventData) => {
+    const file = args.cssFile;
+    if (file === loadedAppCSS) {
+        return;
+    }
+    loadedAppCSS = file;
+
+    loadCss(args.cssFile);
+    application.off("loadAppCss", loadAppCSS);
```
Comment (vakrilov) [inline tns-core-modules/ui/styling/style-scope.ts]: Maybe this is redundant as the `if (file === loadedAppCSS)` will prevent double-execution.

**Comment by vakrilov** [comment]
run ci

File: tns-core-modules/ui/styling/style-scope.ts (hunk 1 above)
Comment (PanayotCankov) [reply tns-core-modules/ui/styling/style-scope.ts]: :+1:

**Comment by PanayotCankov** [comment]
Rebased, should reflect PR comments now.

**Comment by PanayotCankov** [comment]
run ci

**Comment by vakrilov** [comment]
run ci

**Comment by PanayotCankov** [comment]
run ci

**Comment by SvetoslavTsenov** [comment]
run ci

**Comment by SvetoslavTsenov** [comment]
run uitests

**Comment by vchimev** [comment]
ios-pageevents

**Comment by vchimev** [comment]
ios-issues

**Comment by vchimev** [comment]
ios-pickers

**Comment by vchimev** [comment]
ios-textbase

**Comment by hshristov** [comment]
run ci

**Comment by hshristov** [comment]
uitests

**Pull request merged**